translator/internal/internal_to_oc.go

```diff
@@ -0,0 +1,436 @@
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvents = append(ocEvents, eventToOC(event))
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 ocEvents,
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventToOC(event *data.SpanEvent) *octrace.Span_TimeEvent {
+	attrs := event.Attributes()
+
+	ocMessageEventType, ocMessageEventTypeAttrFound := attrs[conventions.OCTimeEventMessageEventType]
+	ocMessageEventID, ocMessageEventIDAttrFound := attrs[conventions.OCTimeEventMessageEventID]
+	ocMessageEventUSize, ocMessageEventUSizeAttrFound := attrs[conventions.OCTimeEventMessageEventUSize]
+	ocMessageEventCSize, ocMessageEventCSizeAttrFound := attrs[conventions.OCTimeEventMessageEventCSize]
+
+	// Consider TimeEvent to be of MessageEvent type if all relevant attributes are set
+	if ocMessageEventTypeAttrFound && ocMessageEventIDAttrFound && ocMessageEventUSizeAttrFound && ocMessageEventCSizeAttrFound {
+		ocMessageEventTypeVal, _ := octrace.Span_TimeEvent_MessageEvent_Type_value[ocMessageEventType.StringVal()]
+		return &octrace.Span_TimeEvent{
+			Time: internal.UnixnanoToTimestamp(uint64(event.Timestamp())),
+			Value: &octrace.Span_TimeEvent_MessageEvent_{
+				MessageEvent: &octrace.Span_TimeEvent_MessageEvent{
+					Type:             octrace.Span_TimeEvent_MessageEvent_Type(ocMessageEventTypeVal),
+					Id:               uint64(ocMessageEventID.IntVal()),
+					UncompressedSize: uint64(ocMessageEventUSize.IntVal()),
+					CompressedSize:   uint64(ocMessageEventCSize.IntVal()),
+				},
+			},
+		}
+	}
+
+	ocAttributes := attributesMapToOCSpanAttributes(attrs, event.DroppedAttributesCount())
+	return &octrace.Span_TimeEvent{
+		Time: internal.UnixnanoToTimestamp(uint64(event.Timestamp())),
+		Value: &octrace.Span_TimeEvent_Annotation_{
+			Annotation: &octrace.Span_TimeEvent_Annotation{
+				Description: &octrace.TruncatableString{
+					Value: event.Name(),
+				},
+				Attributes: ocAttributes,
+			},
+		},
+	}
+}
+
+func linksToOC(links []*data.SpanLink, droppedCount uint32) *octrace.Span_Links {
+	if len(links) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_Links{
+		Link:              linksToOCLinks(links),
+		DroppedLinksCount: int32(droppedCount),
+	}
+}
+
+func linksToOCLinks(links []*data.SpanLink) []*octrace.Span_Link {
+	if len(links) == 0 {
+		return nil
+	}
+
+	ocLinks := make([]*octrace.Span_Link, 0, len(links))
+	for _, link := range links {
+		ocLink := &octrace.Span_Link{
+			TraceId:    link.TraceIDBytes(),
+			SpanId:     link.SpanIDBytes(),
+			Tracestate: traceStateToOC(link.TraceState()),
+			Attributes: attributesMapToOCSpanAttributes(link.Attributes(), link.DroppedAttributesCount()),
+		}
+		ocLinks = append(ocLinks, ocLink)
+	}
+	return ocLinks
+}
+
+func statusToOC(status data.SpanStatus) *octrace.Status {
+	emptyStatus := data.SpanStatus{}
```

Comment:
Can we move this outside the function to ensure it is not created every time (I don't know how well Go's compiler optimizes code like this).

Reply:
done